Index: linux-2.6.23/arch/sh/Makefile
===================================================================
--- linux-2.6.23.orig/arch/sh/Makefile
+++ linux-2.6.23/arch/sh/Makefile
@@ -41,8 +41,13 @@ cflags-$(CONFIG_CPU_SH2)		:= $(call cc-o
 cflags-$(CONFIG_CPU_SH2A)		+= $(call cc-option,-m2a,) \
 					   $(call cc-option,-m2a-nofpu,)
 cflags-$(CONFIG_CPU_SH3)		:= $(call cc-option,-m3,)
+ifdef CONFIG_CPU_SUBTYPE_STX7105
+cflags-$(CONFIG_CPU_SH4)		:= $(call cc-option,-m4-300,) \
+	$(call cc-option,-mno-implicit-fp,-m4-300-nofpu)
+else
 cflags-$(CONFIG_CPU_SH4)		:= $(call cc-option,-m4,) \
 	$(call cc-option,-mno-implicit-fp,-m4-nofpu)
+endif
 cflags-$(CONFIG_CPU_SH4A)		+= $(call cc-option,-m4a,) \
 					   $(call cc-option,-m4a-nofpu,)
 cflags-$(CONFIG_CPU_ST40_300)		+= $(call cc-option,-m4-300,) \
