--- linux-2.6.23-orig/arch/sh/mm/cache-sh4.c	2009-05-18 13:59:28.000000000 +0200
+++ linux-2.6.23/arch/sh/mm/cache-sh4.c	2009-05-18 16:11:21.000000000 +0200
@@ -46,23 +46,23 @@
 
 static void __init emit_cache_params(void)
 {
-	printk("PVR=%08x CVR=%08x PRR=%08x\n",
+	printk(KERN_DEBUG "PVR=%08x CVR=%08x PRR=%08x\n",
 		ctrl_inl(CCN_PVR),
 		ctrl_inl(CCN_CVR),
 		ctrl_inl(CCN_PRR));
-	printk("I-cache : n_ways=%d n_sets=%d way_incr=%d\n",
+	printk(KERN_DEBUG "I-cache : n_ways=%d n_sets=%d way_incr=%d\n",
 		boot_cpu_data.icache.ways,
 		boot_cpu_data.icache.sets,
 		boot_cpu_data.icache.way_incr);
-	printk("I-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
+	printk(KERN_DEBUG "I-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
 		boot_cpu_data.icache.entry_mask,
 		boot_cpu_data.icache.alias_mask,
 		boot_cpu_data.icache.n_aliases);
-	printk("D-cache : n_ways=%d n_sets=%d way_incr=%d\n",
+	printk(KERN_DEBUG "D-cache : n_ways=%d n_sets=%d way_incr=%d\n",
 		boot_cpu_data.dcache.ways,
 		boot_cpu_data.dcache.sets,
 		boot_cpu_data.dcache.way_incr);
-	printk("D-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
+	printk(KERN_DEBUG "D-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
 		boot_cpu_data.dcache.entry_mask,
 		boot_cpu_data.dcache.alias_mask,
 		boot_cpu_data.dcache.n_aliases);
@@ -71,14 +71,14 @@
 	 * Emit Secondary Cache parameters if the CPU has a probed L2.
 	 */
 	if (boot_cpu_data.flags & CPU_HAS_L2_CACHE) {
-		printk("S-cache : n_ways=%d n_sets=%d way_incr=%d\n",
-			boot_cpu_data.scache.ways,
-			boot_cpu_data.scache.sets,
-			boot_cpu_data.scache.way_incr);
-		printk("S-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
-			boot_cpu_data.scache.entry_mask,
-			boot_cpu_data.scache.alias_mask,
-			boot_cpu_data.scache.n_aliases);
+		printk(KERN_DEBUG "S-cache : n_ways=%d n_sets=%d way_incr=%d\n",
+                       boot_cpu_data.scache.ways,
+                       boot_cpu_data.scache.sets,
+                       boot_cpu_data.scache.way_incr);
+		printk(KERN_DEBUG "S-cache : entry_mask=0x%08x alias_mask=0x%08x n_aliases=%d\n",
+                       boot_cpu_data.scache.entry_mask,
+                       boot_cpu_data.scache.alias_mask,
+                       boot_cpu_data.scache.n_aliases);
 	}
 
 	if (!__flush_dcache_segment_fn)
