--- linux-2.6.23/arch/sh/kernel/sh_ksyms.c~	2008-08-06 09:38:31.000000000 +0200
+++ linux-2.6.23/arch/sh/kernel/sh_ksyms.c	2008-08-07 10:56:18.000000000 +0200
@@ -120,7 +120,7 @@
 DECLARE_EXPORT(__movmem_i4_odd);
 DECLARE_EXPORT(__movmemSI12_i4);
 
-#if (__GNUC_MINOR__ == 2 || defined(__GNUC_STM_RELEASE__))
+#if (__GNUC_MINOR__ >= 1 || defined(__GNUC_STM_RELEASE__))
 /*
  * GCC 4.2 emits these for division, as do GCC 4.1.x versions of the ST
  * compiler which include backported patches.
