Index: jfsutils-1.1.12/libfs/devices.c
===================================================================
--- jfsutils-1.1.12.orig/libfs/devices.c
+++ jfsutils-1.1.12/libfs/devices.c
@@ -15,6 +15,9 @@
  *   along with this program;  if not, write to the Free Software
  *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  */
+
+#define _LARGEFILE_SOURCE
+
 #include <config.h>
 #include <errno.h>
 #include <fcntl.h>
